File: vane-regions/src/main/java/org/oddlama/vane/regions/region/RoleSetting.java
package org.oddlama.vane.regions.region;

import org.oddlama.vane.regions.Regions;

public enum RoleSetting {
    ADMIN(false, true),
    BUILD(false, true),
    USE(true, true),
    CONTAINER(false, true),
    PORTAL(false, true);

    private boolean def;
    private boolean defAdmin;

    private RoleSetting(final boolean def, final boolean defAdmin) {
        this.def = def;
        this.defAdmin = defAdmin;
    }

    public boolean defaultValue(final boolean admin) {
        if (admin) {
            return defAdmin;
        }
        return def;
    }

    public boolean hasOverride() {
        return getOverride() != 0;
    }

    public int getOverride() {
        return Regions.roleOverrides.getOverride(this);
    }
}

File: vane-regions/src/main/kotlin/org/oddlama/vane/regions/region/RoleSetting.kt
package org.oddlama.vane.regions.region

import org.oddlama.vane.regions.Regions

/**
 * Permission switches available on region roles.
 */
enum class RoleSetting(private val def: Boolean, private val defAdmin: Boolean) {
    ADMIN(false, true),
    BUILD(false, true),
    USE(true, true),
    CONTAINER(false, true),
    PORTAL(false, true);

    /**
     * Returns the default value for normal users or admins.
     */
    fun defaultValue(admin: Boolean): Boolean = if (admin) defAdmin else def

    /**
     * Returns whether a global override currently applies to this role setting.
     */
    fun hasOverride(): Boolean = override != 0

    /**
     * Global override value from `Regions.roleOverrides`.
     */
    val override: Int
        get() = Regions.roleOverrides?.getOverride(this) ?: 0
}
